File: agents/.github/skills/scaffold-bolton-api-suite-pytest/templates/src/contracts/contract_validator.py
"""Validador de contrato JSON Schema — equivalente Python de
src/api/contracts/contract-validator.ts (que usa AJV no stack TS).
Usa a biblioteca `jsonschema`.
"""

from typing import Any, Dict, List

from jsonschema import Draft7Validator
from jsonschema.exceptions import ValidationError


def compile_contract(schema: Dict[str, Any]) -> Draft7Validator:
    return Draft7Validator(schema)


def format_contract_errors(errors: List[ValidationError]) -> str:
    if not errors:
        return "No contract errors"

    lines = []
    for error in errors:
        path = "/" + "/".join(str(p) for p in error.absolute_path) if error.absolute_path else "/"
        lines.append(f"{path} {error.message}")
    return "\n".join(lines)


def validate_contract(validator: Draft7Validator, instance: Any) -> List[ValidationError]:
    """Retorna a lista de erros (vazia se valido) — chame format_contract_errors()
    para uma mensagem legivel a passar pro assert."""
    return sorted(validator.iter_errors(instance), key=lambda e: list(e.absolute_path))

File: agents/.github/skills/scaffold-bolton-api-suite/templates/src/api/contracts/contract-validator.ts
import Ajv, { type ErrorObject, type ValidateFunction } from 'ajv';
import addFormats from 'ajv-formats';

const ajv = new Ajv({
  allErrors: true,
  strict: true,
});

addFormats(ajv);

export function compileContract(schema: object): ValidateFunction {
  return ajv.compile(schema);
}

export function formatContractErrors(errors: ErrorObject[] | null | undefined): string {
  if (!errors?.length) {
    return 'No contract errors';
  }

  return errors
    .map((error) => `${error.instancePath || '/'} ${error.message ?? 'is invalid'}`)
    .join('\n');
}
